upcloud/resource_upcloud_server.go

```diff
@@ -420,94 +429,112 @@ func resourceUpCloudServerUpdate(ctx context.Context, d *schema.ResourceData, me
-	if d.HasChange("storage_devices") {
-		o, n := d.GetChange("storage_devices")
+		if plan, ok := d.GetOk("plan"); ok {
+			r.Plan = plan.(string)
+		} else {
+			r.CoreNumber = d.Get("cpu").(int)
+			r.MemoryAmount = d.Get("mem").(int)
+			r.Plan = "custom"
+		}
 
-		// detach the devices that should be detached or should be re-attached with different parameters
-		for _, storageDevice := range o.(*schema.Set).Difference(n.(*schema.Set)).List() {
-			if serverDetails.StorageDevice(storageDevice.(map[string]interface{})["storage"].(string)) == nil {
-				continue
+		// handle the template
+		if d.HasChanges("template.0.title", "template.0.size", "template.0.backup_rule") {
+			template := d.Get("template.0").(map[string]interface{})
+			if _, err := client.ModifyStorage(&request.ModifyStorageRequest{
+				UUID:  template["id"].(string),
+				Size:  template["size"].(int),
+				Title: template["title"].(string),
+				BackupRule: storage.BackupRule(
+					d.Get("template.0.backup_rule.0").(map[string]interface{}),
+				),
+			}); err != nil {
+				return diag.FromErr(err)
 			}
+		}
+		// should reattach if address changed
+		if d.HasChange("template.0.address") {
+			o, n := d.GetChange("template.0.address")
 			if _, err := client.DetachStorage(&request.DetachStorageRequest{
 				ServerUUID: d.Id(),
-				Address:    storageDevice.(map[string]interface{})["address"].(string),
+				Address:    o.(string),
 			}); err != nil {
 				return diag.FromErr(err)
 			}
-		}
-		// attach the storages that are new or have changed
-		for _, storageDevice := range n.(*schema.Set).Difference(o.(*schema.Set)).List() {
-			storageDevice := storageDevice.(map[string]interface{})
 			if _, err := client.AttachStorage(&request.AttachStorageRequest{
+				Address:     n.(string),
 				ServerUUID:  d.Id(),
-				Address:     storageDevice["address"].(string),
-				StorageUUID: storageDevice["storage"].(string),
-				Type:        storageDevice["type"].(string),
+				StorageUUID: d.Get("template.0.id").(string),
 			}); err != nil {
 				return diag.FromErr(err)
 			}
 		}
+
+		// handle the other storage devices
+		if d.HasChange("storage_devices") {
+			o, n := d.GetChange("storage_devices")
+
+			// detach the devices that should be detached or should be re-attached with different parameters
+			for _, storageDevice := range o.(*schema.Set).Difference(n.(*schema.Set)).List() {
+				deviceUUID := storageDevice.(map[string]interface{})["storage"].(string)
+				if serverDetails.StorageDevice(deviceUUID) == nil {
+					continue
+				}
+				if _, err := client.DetachStorage(&request.DetachStorageRequest{
+					ServerUUID: d.Id(),
+					Address:    storageDevice.(map[string]interface{})["address"].(string),
+				}); err != nil {
+					return diag.FromErr(err)
+				}
+			}
+			// attach the storages that are new or have changed
+			for _, storageDevice := range n.(*schema.Set).Difference(o.(*schema.Set)).List() {
+				storageDevice := storageDevice.(map[string]interface{})
```

Comment:
I like this approach better than the one above - eg. get the item from the list and cast it once and then pick the relevant attributes out of it.

However, using the same name here could be a bit dangerous due to the scope being maybe a bit confusing? Maybe something like;
```
	for _, rawStorageDevice := range n.(*schema.Set).Difference(o.(*schema.Set)).List() {
				storageDevice := rawStorageDevice.(map[string]interface{})
``` 
or, changedStorageDeviceData or something? Or; even better, make a helper that extracts (casted) added/removed lists from the changes? So something along the lines of;
```
func getStorageDeviceChanges(d *schema.ResourceData) (removedDevices []map[string]interface{}, addedDevices []map[string]interface{}) {
... 
}
``` 

But tbh, it's probably just me being grumpy in the afternoon - this works for me too if you'd rather jump on to something else, up to you!

Reply:
this seems to be more readable, I'll do it

Reply:
The issue is this way was implemented all over the codebase :disappointed: